funfact/quantum/_qgates.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from abc import ABC, abstractmethod
+from numbers import Integral
+from typing import List
+from funfact import active_backend as ab
+import numpy as np
+from funfact.parametrized import Generator
+from funfact.lang._tsrex import TsrEx, tensor
+from funfact.lang._ast import Primitives as P
+from funfact.lang._terminal import ParametrizedAbstractTensor
+from funfact.lang._special import eye
+from ._special import proj0, proj1
+from funfact.conditions import Unitary
+# TODO: how to handle dtype? Should always be complex!
+
+
+class Gate(ABC):
+    '''A common base class for all gate objects.'''
+    def __init__(self, qubits: List[Integral]):
+        '''A gate stores the minimum and maximum qubits it acts on.'''
+        for i, q in enumerate(qubits):
+            if not (isinstance(q, Integral) and q >= 0):
+                raise RuntimeError(
+                    f"Qubit indices must be non-negative integers, got {q} for"
+                    f" entry {i}."
+                )
+        qubits.sort()
```

Comment:
`.sort()` modifies the input argument in-place and may cause unexpected behavior in a complex program. A safer option is to use `sorted` and store the result as a `tuple`, which is immutable.